banzaicloud/pipeline
# use latest velero chart & API and prefix to buckets for cluster backup

| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | fixes #3215 
| License         | Apache 2.0


### What's in this PR?
- use latest Velero 1.5.1 API, library & helm chart modified slightly to support different cloud provider secrets at a time
- add prefix field to ark_backup_buckets table
- set cluster name as a prefix for buckets when creating backups

### Why?
To be able to support using same object store bucket for multiple clusters

### Checklist


- [X] Implementation tested (with at least one cloud provider)
- [X] Code meets the [Developer Guide](https://github.com/banzaicloud/developer-guide)
- [X] OpenAPI and Postman files updated (if needed)
- [X] User guide and development docs updated (if needed)
- [X] Related Helm chart(s) updated (if needed)

### Test Plan

This PR has been tested following these steps:

- create EKS/PKE AWS cluster
- `helm install wordpress  bitnami/wordpress`
- create a new blog record
- create backup to s3 new bucket
- check backup is created in bucket
- create new EKS cluster from the previously created backup
- check the previously create blog record is there, while Velero resources were deleted
- create backup to GS bucket, check backup dir is created on Google Console 
- delete backup, create backup to Azure Blob Storage, check backup is created on Azure Console

- create GKE cluster, create backup to GS and S3 then restore new cluster from backup
- create AKS/PKE Azure cluster, create backup to Azure Blob Storage and S3 then restore new cluster from backup

### Upgrade plan

Following the upgrade, Pipeline will attempt to sync the followings:

- buckets from objectstore: since Velero backup format is different then earlier Ark format, no backups will be found in storage buckets, all existing backups going to be deleted from Pipeline database.
- backup and restore resources from cluster: as the new custom resources group (velero.io) is different then the CRD group installed on clusters (ark.io) there will be no backup/restore found on cluster

In short: earlier backups will be deleted from Pipeline database (backup resources on cluster, backup files on object storage buckets stay untouched) scheduled backups are performed by Ark operator running on clusters, however users are not able to do backups/restores with Pipeline UI/CLI until backup is disabled and enabled again, installing the latest Velero operator.

Users can disable backup using banzai cli:

```
banzai cluster --cluster-name=CLUSTER_NAME service backup disable
```
